core/drivers/zynqmp_pm.c

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2021 Foundries.io Ltd
+ */
+
+#include <arm.h>
+#include <drivers/zynqmp_pm.h>
+#include <kernel/cache_helpers.h>
+#include <mm/core_memprot.h>
+#include <smccc.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+#include <utee_defines.h>
+
+#define EFUSE_ACCESS_SMC	0xC2000035
+#define EFUSE_NOT_ENABLED	29
+
+enum efuse_op {EFUSE_READ = 0, EFUSE_WRITE = 1};
+
+/*
+ * Stores all required details to
+ * read/write efuse memory.
+ * @src:        paddr_t of the buffer to store the data to be write/read
+ *              buffer should be cache aligned
+ * @size:       number of 32-bit words to be read/written
+ * @offset:     offset to be read/written
+ * @flag:       EFUSE_READ  - represents efuse read operation
+ *              EFUSE_WRITE - represents efuse write operation
+ * @pufuserfuse:0 - represents non-puf efuses, offset is used for read/write
+ *              1 - represents puf user fuse row number.
+ */
+struct xilinx_efuse {
+	uint64_t src;
+	uint32_t size;
+	uint32_t offset;
+	uint32_t flag;
+	uint32_t pufuserfuse;
+};
+
+static uint32_t zynqmp_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				uint32_t *payload)
+{
+	struct smccc_res res = { };
+
+	smccc_smc(pm_api_id,
+		  ((uint64_t)arg1 << 32) | arg0,
+		  ((uint64_t)arg3 << 32) | arg2,
+		  0, 0, 0, 0, 0, &res);
+
+	if (payload && !res.a0)
+		*payload = res.a0 >> 32;
```

Comment:
No need to shift, `res.a0` is always `0` if we reach here.

Reply:
that wasn't correct at all, I've checked how it's handled in the linux kernel and corrected
res.a0 contains both return code and payload